sklearn/metrics/cluster/tests/test_unsupervised.py

```diff
@@ -170,6 +170,22 @@ def test_non_numpy_labels():
         silhouette_score(list(X), list(y)), silhouette_score(X, y))
 
 
+def test_silhouette_nonzero_diag():
+    # Construct a zero-diagonal matrix
+    dists = pairwise_distances(
+        np.array([[0.2, 0.1, 0.12, 1.34, 1.11, 1.6]]).transpose())
+
+    # Construct a nonzero-diagonal distance matrix
+    diag_dists = dists.copy()
+    np.fill_diagonal(diag_dists, 1)
+
+    labels = [0, 0, 0, 1, 1, 1]
+
+    assert_raise_message(ValueError, "distance matrix contains non-zero",
```

Comment:
we use `with pytest.raises(ValueError, match=...):` now, but I wouldn't prevent this from merging.